enhancements/agent-installer/automated-workflow.md

```diff
@@ -0,0 +1,504 @@
+monitoring will need to shift to the OpenShift APIs (`ClusterVersion`
+and `ClusterOperator`). We will extend the `openshift-install wait-for
+install-complete` command to recognize when the cluster being deployed
+uses the agent installer and to use both APIs for monitoring and
+[progress
+reporting](https://github.com/openshift/assisted-service/blob/master/docs/enhancements/installation-progress-bar.md)
+in that case, including reporting when steps like validation fail and
+the cluster deployment is never going to succeed. Users will still be
+able to write their own monitoring logic using the assisted service
+and OpenShift APIs directly, but they will have the option of using
+the command line tool.
+
+For the fully automated workflow, the steps a user might normally
+perform such as approving hosts or triggering the installation will be
+handled by a new client running as a service on node0. The client will
+read the user's inputs from files included in the image and interact
+with the assisted-service REST API on localhost to drive the
+deployment. For the purposes of this document, we will refer to that
+service as a single entity called `create-cluster`, although in
+practice we may divide some of the steps into separate tools.
+
+### Actors in the Workflow
+
+**infrastructure admin** is a human user responsible for infrastructure
+within the data center, such as an image registry, networking, etc.
+
+**cluster creator** is a human user responsible for deploying a cluster.
+
+**orchestration tool** is intended to be a software component provided
+by the end user, but the steps may also be performed by **cluster
+creator** if no software component is available.
+
+**installer** is the OpenShift installer binary, `openshift-install`.
+
+**node0** is the first host to be added to a cluster.
+
+### Steps for Deploying a Disconnected Cluster
+
+1. The infrastructure admin launches an image registry visible to the
+   hosts that will form the cluster and to a host that has internet
+   access so the release image can be downloaded.
+2. The cluster creator obtains the `oc` command line tool and uses `oc`
+   to copy the OpenShift release image into the local image registry
+   in one of the usual ways with [`oc adm
+   mirror`](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-installation-images.html)
+   or [the `oc mirror`
+   plugin](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-disconnected.html).
+3. The cluster creator obtains the OpenShift installer in one of [the
+   usual
+   ways](https://docs.openshift.com/container-platform/4.9/installing/installing_bare_metal/installing-bare-metal.html#installation-obtaining-installer_installing-bare-metal).
+4. The cluster creator triggers their orchestration tool to run the
+   installer.
+5. The orchestration tool prepares an `install-config.yaml`, including
```

Comment:
I thought I wrote a comment about install-config yesterday, but it seems to have disappeared.  This is the only place in the document where install-config.yaml is mentioned, and it's not clear what it's for.  Is the input to this whole process going to be the install-config.yaml, and not ZTP manifests?  Are we going to turn install-config into ZTP manifests, which in turn get turned into install-config?

Reply:
installConfig, to me, is optional in this step .If you provide it, it will be used to generate content for the ZTP manifests, which are then put into the ISO. If you provide ZTP manifests from the get go, no installconfig should be necessary, as it is later generated by assisted-service.

Reply:
What's the point of having two APIs to the same tool?

Reply:
AI generates install-config. What will happen if it will disagree with the provided one?
We discussed it in the past while providing hive apis and finally decided to allow only patches in order to avoid the bad ux and complexity of merging. Not sure I understand the reasoning to have it as the main input - it already lacks some of the functionality, so having ztp manifests is not something that we'll be able to avoid.

Reply:
It's an open question (that's what https://github.com/openshift/enhancements/pull/1067#discussion_r831557200 is referring to).

> Are we going to turn install-config into ZTP manifests, which in turn get turned into install-config?

This is definitely an option we're considering.

Reply:
If we use install-config as the input, I don't really see the benefit of converting to ZTP APIs as an intermediate step. We've established that the asset graph management code in the installer doesn't let the user modify the manifests, only add to them. So, we could simply copy the user-provided install-config into the image, then read that to drive the REST API. The fact that the assisted-service is going to generate a different install-config based on those API calls so it can run the installer binary again is an implementation detail.

Reply:
> We've established that the asset graph management code in the installer doesn't let the user modify the manifests, only add to them.

Did we establish that? I thought what we established was:

1. Both install-config and manifests present; install-config modified -> does the Right Thing
2. Both install-config and manifests present; _both_ modified -> manifest changes are overwritten(!)
3. Both install-config and manifests present; manifests modified -> modified manifests used(???)
4. Only manifests present: not possible today, but we can hack it so that the install-config is optional

If (3) doesn't work like that then I agree that would be a compelling reason to only use only the install-config and not have ZTP manifests in a create-manifests stage.

Reply:
What I heard @staebler say was that if you have an install-config then the installer will generate fresh copies of the other manifests in the dependency graph every time it runs. I see that `create manifests` removes the install-config input file, so it wouldn't be there the second time around, and so we could potentially do something similar. I'm still not sure I see the point, though. Do we expect the ZTP APIs to support features that aren't supported in the install-config? Why would we want to recommend to the user that they modify any intermediate files?

Reply:
If ZTP manifest assets are dependencies of the install config asset, then the installer will consider the ZTP manifest assets dirty if the install-config.yaml file is present. As a consequence, the installer will re-generate the ZTP manifest assets, overwriting any changes made by the user to any asset files that the installer generates.

Reply:
This was resolved, but I still didn't see any compelling reason to not use the ZTP API as originally discussed.